src/axom/multimat/docs/sphinx/multimat_using_fields.rst

```diff
@@ -0,0 +1,308 @@
+.. ## Copyright (c) 2017-2024, Lawrence Livermore National Security, LLC and
+.. ## other Axom Project Developers. See the top-level LICENSE file for details.
+.. ##
+.. ## SPDX-License-Identifier: (BSD-3-Clause)
+
+******************************************************
+Using Fields
+******************************************************
+
+Fields store useful simulation quantities. MultiMat supports defining fields
+on the mesh and material subsets of the mesh. This section describes how to
+create fields and access their data.
+
+#######################
+Adding a Field
+#######################
+
+The ``addField()`` method adds a field to a MultiMat object. The method
+accepts arguments that indicate the mapping, layout, and sparsity for the supplied
+data, which are given using an ``axom::ArrayView``. The data given in the view are
+copied into new memory managed by MultiMat.
+
+The field mapping argument indicates the space where the data live: the mesh cells,
+the materials, or the cells/material regions defined over the mesh. The
+data layout argument indicates how the data are organized with respect to cells and
+materials. For data that have 1 value per cell, pass ``PER_CELL``. For data that have
+1 value per material *(ignoring how many cells use the material)*, pass ``PER_MAT``. For
+data that have a unique value per material within a cell, pass ``PER_CELL_MAT``.
+For ``PER_CELL_MAT`` data, it is important to know the data layout. Pass ``CELL_DOM``
+if all of the material values for a cell are sequential neighbors in memory; otherwise
+pass ``MAT_DOM``.
+
+Sparsity layout indicates whether the data array contains the maximum number of
+values (numMaterials * numCells for ``DENSE`` fields) or whether it instead contains
+only the subset of elements where materials are defined. The length of ``SPARSE``
+fields is determined by the number of true values in the Cell-Mesh Relation (CMR).
+
+.. code-block:: cpp
+
+  constexpr int nComponents = 1;
+  double data[] = {0.,
+                   1.,
+                   2.,
+                   // more values ...
+                  };
+  axom::ArrayView<double> dataAV(data, sizeof(data) / sizeof(double));
+  mm.addField("x2",
+              axom::multimat::FieldMapping::PER_CELL,
+              axom::multimat::DataLayout::CELL_DOM,
+              axom::multimat::SparsityLayout::SPARSE,
+              dataAV,
+              nComponents);
```

Comment:
It would be very helpful to me if you could expand this example to have a diagram of (say) a 2x3 mesh with maybe two materials.  Perhaps you could skip the call to setCellMatRel() and setVolfracField().  But it would help if you had three calls to addField(), one with ``PER_CELL``,  one with ``PER_MAT``, and one with ``PER_CELL_MAT`` and ``CELL_DOM``.  These could all be DENSE, because you haven't gotten to the discussion of index sets yet.  I think seeing everything fully written out would help readers learn these concepts.

Reply:
I moved code snippets into a basic.cpp example that compiles and added more examples of adding fields.